src/wallet/spend.cpp

```diff
@@ -373,76 +373,72 @@ std::vector<OutputGroup> GroupOutputs(const CWallet& wallet, const std::vector<C
     }
 
     // The knapsack solver has some legacy behavior where it will spend dust outputs. We retain this behavior, so don't filter for positive only here.
     std::vector<OutputGroup> all_groups = GroupOutputs(wallet, coins, coin_selection_params, eligibility_filter, false /* positive_only */);
     // While nTargetValue includes the transaction fees for non-input things, it does not include the fee for creating a change output.
     // So we need to include that for KnapsackSolver as well, as we are expecting to create a change output.
-    std::set<CInputCoin> knapsack_coins;
-    CAmount knapsack_value;
-    if (KnapsackSolver(nTargetValue + coin_selection_params.m_change_fee, all_groups, knapsack_coins, knapsack_value)) {
-        const auto waste = GetSelectionWaste(knapsack_coins, coin_selection_params.m_cost_of_change, nTargetValue + coin_selection_params.m_change_fee, !coin_selection_params.m_subtract_fee_outputs);
-        results.emplace_back(std::make_tuple(waste, std::move(knapsack_coins), knapsack_value));
+    std::optional<SelectionResult> knapsack_result = KnapsackSolver(all_groups, nTargetValue + coin_selection_params.m_change_fee);
+    if (knapsack_result) {
+        results.push_back(std::make_pair(*knapsack_result, knapsack_result->GetWaste(coin_selection_params.m_cost_of_change)));
     }
 
     // We include the minimum final change for SRD as we do want to avoid making really small change.
     // KnapsackSolver does not need this because it includes MIN_CHANGE internally.
     const CAmount srd_target = nTargetValue + coin_selection_params.m_change_fee + MIN_FINAL_CHANGE;
     auto srd_result = SelectCoinsSRD(positive_groups, srd_target);
     if (srd_result != std::nullopt) {
-        const auto waste = GetSelectionWaste(srd_result->first, coin_selection_params.m_cost_of_change, srd_target, !coin_selection_params.m_subtract_fee_outputs);
-        results.emplace_back(std::make_tuple(waste, std::move(srd_result->first), srd_result->second));
+        results.push_back(std::make_pair(*srd_result, srd_result->GetWaste(coin_selection_params.m_cost_of_change)));
     }
 
     if (results.size() == 0) {
         // No solution found
-        return false;
+        return std::nullopt;
     }
 
     // Choose the result with the least waste
     // If the waste is the same, choose the one which spends more inputs.
-    const auto& best_result = std::min_element(results.begin(), results.end(), [](const auto& a, const auto& b) {
-        return std::get<0>(a) < std::get<0>(b) || (std::get<0>(a) == std::get<0>(b) && std::get<1>(a).size() > std::get<1>(b).size());
+    auto [best_result, waste] = *std::min_element(results.begin(), results.end(), [](const auto& a, const auto& b){
+        const auto [a_result, a_waste] = a;
+        const auto [b_result, b_waste] = b;
+        return a_waste < b_waste || (a_waste == b_waste && a_result.m_selected_inputs.size() > b_result.m_selected_inputs.size());
```

Comment:
Can this be written into a comparison operator for `SelectionResult`? The logic of waste and what makes one selection result better than another is inherent to the `SelectionResult` type; imo it fits better there than within `AttemptSelection`. 

If you have a `<` operator defined, you can sort and/or just return the min element.

Reply:
I didn't want to always be computing waste every time, especially because we may not have access to information such as the cost of change. So the approach I've done for this is to have a function which computes and stores the waste inside of `SelectionResult`, and then `operator<` will use that when doing the comparison.